### ISSUE TITLE
onMounted hook running before DOM is available

### ISSUE THREAD

**Issue by fanckush** [nuxt/framework#3587]
### Environment

- Operating System: `Darwin`
- Node Version:     `v17.3.1`
- Nuxt Version:     `3.0.0-27447229.20f3171`
- Package Manager:  `yarn@1.22.17`
- Bundler:          `Vite`
- User Config:      `vite`, `build`, `ssr`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

Not needed. The Bug is very clear

### Describe the bug

The screenshot below says it all. On first load (refresh) everything is fine and I'm able to run `querySelector` and access the DOM just fine. However if I navigate to a different route then come back I get `null`.
``` vue
<template>
  <div class="TheAdvantages">
    <ul>
      <li class="first" @click="$router.push('')"> I'm first </li>
    </ul>
  </div>
</template>

<script setup lang="ts">
onMounted(() => {
  console.log('onMonuted Hook:')

  console.log(
    "document.querySelector('.first'): ",
    document.querySelector('.first')
  )

  setTimeout(() => {
    console.log('setTimeout 0ms')
    console.log(
      "document.querySelector('.first'): ",
      document.querySelector('.first')
    )
  }, 0)

})
```

<img width="584" alt="image" src="https://user-images.githubusercontent.com/6278713/157617245-898d6400-b18b-4ed9-9f27-cc90883d82a5.png">

**Comment by fanckush** [comment]
For anyone else having this issue until it's resolved you can use `setTimeout` 

Instead of:

``` js
onMounted(() => {
... code that needs DOM // WON'T WORK
})
```

Do this:
``` js
onMounted(() => {
  setTimeout(() => {
    ... code that needs DOM // WORKS!! 
  }),
    0
})

```

**Comment by bcspragu** [comment]
I just ran into the same thing, working with the PayPal SDK. Hard refreshes work fine, and soft navigation works fine as well, but the second navigation calls `onMounted` before the page is rendered. When I drop a `debugger` in, I can see I'm still clearly on the previous page (where I clicked the link), and so the target element doesn't exist.

**Comment by fanckush** [comment]
@the-m-git could you provide a snippet?

**Comment by danielrjames** [comment]
I had a similar issue where I could not call .focus() on a template ref in onMounted(). I found that the setTimeout hack does work, but it's pretty messy, so I'm not really a fan of that.

However, I came across https://github.com/nuxt/framework/issues/2801 and found that if you set `<NuxtPage :key="$route.fullPath" />` in you `app.vue` file, it works as you would expect, so no need to setTimeout, etc.

I have no idea why this would be the solution. @danielroe thoughts?

**Comment by fanckush** [comment]
@danielrjames thanks! interesting workaround
The `key` forces the component to be destroyed and re-rendered on each route change.
I suppose this fixes the issue by forcing a full lifecycle of the page component which eventually leads to a proper `mounted` hook, I'll give it a try.

I imagine this to affect performance, because vue won't get to reuse anything really... so while this might be more stable than setTimeout it's certainly still just a workaround

**Comment by danielrjames** [comment]
> ... so while this might be more stable than setTimeout it's certainly still just a workaround

I agree. I do think the advantage of setting the key on `NuxtPage` over using `setTimeout` in `onMounted` is that when the nuxt team finds time to fully investigate and implement a fix for this bug, you won't have to update your components.

**Comment by danielroe** [comment]
It seems this is in fact a vue core bug: https://github.com/vuejs/core/issues/5844.

**Comment by danielrjames** [comment]
@danielroe is there a way to disable transitions globally?

**Comment by fanckush** [comment]
### Workaround - Disable Transitions 

in every page where you care about onMounted running accurately add this inside `setup`:
``` js
definePageMeta({
  pageTransition: false,
}) 
```

notes:
- I haven't found a way to globally disabled transitions 
- I'm not sure on what i'm missing out by disabling the transition

**Comment by nndnha** [comment]
> ### Workaround - Disable Transitions
> in every page where you care about onMounted running accurately add this inside `setup`:
> 
> ```js
> definePageMeta({
>   pageTransition: false,
> }) 
> ```
> 
> notes:
> 
> * I haven't found a way to globally disable transitions
> * I'm not sure on what i'm missing out by disabling the transition

It works but in my case I got another issue on doing as you instructed https://github.com/vuejs/core/issues/3989

**Comment by fanckush** [comment]
@nndnha do you're pages have a single root element?

**Comment by ironexdev** [comment]
> ### Workaround - Disable Transitions
> in every page where you care about onMounted running accurately add this inside `setup`:
> 
> ```js
> definePageMeta({
>   pageTransition: false,
> }) 
> ```
> 
> notes:
> 
> * I haven't found a way to globally disable transitions
> * I'm not sure on what i'm missing out by disabling the transition

This seems to be working for me (globally / per layout)

```
<NuxtPage :transition="false" />
```

**Comment by danielroe** [comment]
You can globally disable transitions like this:

```js
export default defineNuxtConfig({
  app: {
    pageTransition: false,
    layoutTransition: false,
  }
})
```

**Comment by treading0light** [comment]
> ### Workaround - Disable Transitions
> in every page where you care about onMounted running accurately add this inside `setup`:
> 
> ```js
> definePageMeta({
>   pageTransition: false,
> }) 
> ```
> 
> notes:
> 
> * I haven't found a way to globally disable transitions
> * I'm not sure on what i'm missing out by disabling the transition

Thank you Thank you, the main fix on this thread didn't fix my issue but this did.

**Comment by richgcook** [comment]
> I actually increased my setTimeout to 100 and that seemed to be long enough to load the dom.

Hard to understand we have to rely on this...

**Comment by victorz** [comment]
> I actually increased my setTimeout to 100 and that seemed to be long enough to load the dom.

Do note that what works for your machine, timing wise, might not work for other machines, depending on speed, load, all kinds of things.

**Comment by richgcook** [comment]
> You can globally disable transitions like this:
> 
> ```js
> export default defineNuxtConfig({
>   app: {
>     pageTransition: false,
>     layoutTransition: false,
>   }
> })
> ```

Did this come in RC12? This wasn't possible until recently (you previously had to set it on every page using `definePageMeta`)

**Comment by manniL** [comment]
Yes. In the next rc, transitions will be disabled by default - #8436

**Comment by michaelpumo** [comment]
Unfortunately, I don't think disabling transitions is a sufficient "work-around" if your client and designers are expecting to see transitions. Was this a Vue bug?

I can get the DOM element in `onMounted()` but what I can't do is read its dimensions. So whilst the element is _not_ `null`, it returns `0` when running `element.getBoundingClientRect().width` for example.

**Comment by danielroe** [comment]
Yes, it is an upstream Vue bug, linked above. You might be able to await nextTick to workaround it.

**Comment by nandi95** [comment]
You could work around it by doing logic in the `page:transition:finish` hook

**Comment by fskarmeta** [comment]
good that i found this :D , also having to use nextTick now since rc-12 to load my leaflet code on my client components onMounted hook.

**Comment by scanderoff** [comment]
> good that i found this :D , also having to use nextTick now since rc-12 to load my leaflet code on my client components onMounted hook.

nextTick is not working

**Comment by therealemino** [comment]
I just ran into a somewhat similar issue,
When I navigate to a different route and come back, everything works fine. However on fresh load, it returns null.

This only happens when I go live. 
If I run "nuxt start", everything still works fine.

**Comment by benjaminrobinet** [comment]
When using `key` attribute on `NuxtPage`, the transition is not even triggered ?
Do anyone having any fix to solve that ?

**Comment by pslhrd** [comment]
`    <Transition name="appear" mode="out-in">
      <div :key="$route.fullPath"> 
        <NuxtPage />
      </div>
    </Transition>`

Thats how to keep your transition // and having dom properly loaded, seems weird with all the caching going on its not optimal.

**Comment by richgcook** [comment]
Even if you add `pageTransition: false` and `layoutTransition: false` to config the DOM still isn't retrievable when using, for example, `elem.value.offsetLeft`, in `onMounted`...

**Comment by ElderlyBoy** [comment]
> You could work around it by doing logic in the `page:transition:finish` hook

but `page:transition:finish` is running globally

**Comment by ElderlyBoy** [comment]
I meet the problem like this:
````vue
<script lang="ts" setup>
  import {getMockData} from '@/apis'
  //Fetch Data
  const {data: mockData } = await getMockData();
  //some code for mockData
  //...
  //needs to get DOM
  nextTick(() => {
    console.log(document.querySelector('.exist-class'));
    //ssr out: <div class=".........
    //csr out: null
  })
  //same in mounted lifecycle hook
  onMounted(() => {
    console.log(document.querySelector('.exist-class'));
    //ssr out: <div class=".........
    //csr out: null
  })
</script>
````

**Comment by cyrenity** [comment]
If I enable page transitions, this piece of code responsible for loading Stripe elements inside my component stops working. Disabling page transitions fixes the issue.


`onMounted(async () => {
  const ELEMENT_TYPE = 'card'
  stripe = await loadStripe(stripeKey.value.key)

  elements = stripe.elements()
  const element = elements.create(ELEMENT_TYPE, style)
  element.mount('#card-element')
})`
